### PR TITLE
Pipeline to device

### LINKED ISSUE #195: Implement `pipeline.to(device)`
Currently, pipeline modules are moved to the preferred compute device during `__call__`. This is reasonable, as they stay there as long as the user keeps passing the same `torch_device` across calls.

However, in multi-GPU model-serving scenarios, it could be useful to move each pipeline to a dedicated device during or immediately after instantiation. This would make it possible to create, say, 8 different pipelines and move each one to a different GPU. Doing it this way could potentially save CPU memory while preparing the service.

Currently, the workaround to achieve the same would be to perform a call with fake data immediately after the instantiation.

**Describe the solution you'd like**
Ideally, the following should work:
```Python
pipe = StableDiffusionPipeline.from_pretrained(model_id).to("cuda:1")
```

**Describe alternatives you've considered**
Current workaround:
```Python
pipe = StableDiffusionPipeline.from_pretrained(model_id)
_ = pipe(["cat"], num_inference_steps=1, torch_device="cuda:1")
```

Another alternative would be to pass the device to the initializer. This could be done in addition to adding a `to` method, but I believe it's not necessary as `to` is familiar enough to PyTorch users.

**Additional context**
See discussion in [this Slack thread](https://huggingface.slack.com/archives/C03HBN1C8CW/p1660716555104789).

### PR DESCRIPTION
**Pull request by pcuenca**
The following will run the pipeline in `cuda:1` (not `cuda:0`) as would be expected:

```Python
pipe = StableDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-3-diffusers",
    use_auth_token=True).to("cuda:1")
pipe("Some prompt")
```

However, passing a device to `__call__` will still move the pipeline to that device.

The implementation of `DiffusionPipeline.to()` does nothing if the device is `None`. This is to preserve the same semantics as PyTorch, where AFAIK if you use `to(None)` the object is not moved anywhere.

If we were to forgo that behaviour, we could make `DiffusionPipeline.to(None)` select `cuda` by default (when available). This would make for much simpler code in all pipeline implementations, as they'd just need to invoke `self.to`, but might break the expectations of PyTorch users. The current implementation in all the pipelines `__call__` methods does exactly that: select `cuda` if available, and if no previous device was already set by the user. It is a bit repetitive and maybe a bit fragile.

Is it important to preserve PyTorch semantics in this regard, or is it better to make all `__call__` implementations simpler? What do you think @anton-l @patil-suraj @patrickvonplaten ?

Fixes #195

### CODE REVIEW

**Review by anton-l — CHANGES_REQUESTED** [review]
Thanks for the PR @pcuenca!

Haven't seen cases of `.to(None)` in the wild before, so to me it's fine to break the expectations a bit :) 
So implementing it like this:
```python
def to(self, torch_device: Optional[Union[str, torch.device]] = None):
        if torch_device is None:
            torch_device = "cuda" if torch.cuda.is_available() else "cpu"
        ...
```

And then just doing `self.to(torch_device)` inside the pipelines would be much cleaner and less implementation error-prone for future pipelines

**Comment by pcuenca** [comment]
> Haven't seen cases of `.to(None)` in the wild before, so to me it's fine to break the expectations a bit :) 

Cool, I agree! That's how I would have done it if it had been for an internal project. Because it's open source, I might have been extremely cautious here :)

> 
> ```python
> def to(self, torch_device: Optional[Union[str, torch.device]] = None):
>         if torch_device is None:
>             torch_device = "cuda" if torch.cuda.is_available() else "cpu"
>         ...
> ```
> 
> And then just doing `self.to(torch_device)` inside the pipelines would be much cleaner and less implementation error-prone for future pipelines

One thing, though. If we expect users to use `.to(device)` and then _not_ provide any device during `__call__`, our implementation must avoid moving the models to the default cuda device if a previous one was selected.

I'll prepare it right away so you can take a look.

**Comment by pcuenca** [comment]
@anton-l not super happy after the change, `.to()` works as expected, but the `__call__` methods need to be careful to use `self.device` instead of the argument. What do you think?

**Comment by anton-l** [comment]
Hmm, indeed. But choosing between the [first version](https://github.com/huggingface/diffusers/pull/210/commits/c5c9a6f5b74d1b4b353e84de1ad58db9ddfb5d6f) and this one, I would still go with the current `image.to(self.device)` :slightly_smiling_face: 
Guess there are just two options available, wdyt @patil-suraj?

**Review by patrickvonplaten — COMMENTED** [review]
Throwing in an idea -> we could also just remove `torch_device` as in input to the forward function. It's more "PyTorch'y" to use the `.to(...)` API IMO. 

So we could just deprecate the `"torch_device"` parameter, saying that it'll be removed in `0.3.0` and only rely on `to(...)`.

What do you think?

**Comment by pcuenca** [comment]
> we could also just remove `torch_device` as in input to the forward function.

I think that's a very sensible idea. But I also love that if you do nothing it woks as expected (uses cuda if available), so we'd do an automatic placement on creation. Does that sound right?

File: src/diffusers/pipeline_utils.py [hunk 1]
```diff
@@ -112,6 +113,31 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike]):
             save_method = getattr(sub_model, save_method_name)
             save_method(os.path.join(save_directory, pipeline_component_name))
 
+
+    def to(self, torch_device: Optional[Union[str, torch.device]] = None):
+        modules, _ = self.extract_init_dict(dict(self.config))
```
Comment (patil-suraj) [inline src/diffusers/pipeline_utils.py:118]: Let's maybe call this `module_names` as these are not actual modules.
```suggestion
        module_names, _ = self.extract_init_dict(dict(self.config))
```

File: src/diffusers/pipeline_utils.py [hunk 2]
```diff
@@ -112,6 +113,31 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike]):
...
+        for name in modules.keys():
+            module = getattr(self, name)
+            if isinstance(module, torch.nn.Module):
+                if torch_device is None:
+                    if module.device.type == "cpu":
+                        torch_device = "cuda" if torch.cuda.is_available() else "cpu"
+                    else:
+                        torch_device = module.device
```
Comment (patil-suraj) [inline src/diffusers/pipeline_utils.py:126]: I'm actually in favor of not doing anything when `torch_device`. Think it's better to have same semantics as the pt `.to`.

File: src/diffusers/pipeline_utils.py [hunk 3]
```diff
@@ -112,6 +113,31 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike]):
...
+                module.to(torch_device)
+        return self
+
+
+    @property
+    def device(self) -> torch.device:
+        modules, _ = self.extract_init_dict(dict(self.config))
```
Comment (patil-suraj) [inline src/diffusers/pipeline_utils.py:133]: same comment as above.

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR @pcuenca !

I agree with @patrickvonplaten about removing the `torch_device` argument. Would be nice to have just one API for device handling, rather than having both options. That would be much cleaner IMO.

**Comment by patrickvonplaten** [comment]
> torch_device

I'd actually say to not do an automatic displacement to really stay 1-to-1 the same as PyTorch. I really like the fact that you know in PyTorch models are always on "cpu" by default. 

@anton-l @patil-suraj what do you think?

File: src/diffusers/pipeline_utils.py (hunk 1 above)
Comment (pcuenca) [reply src/diffusers/pipeline_utils.py:118]: Good call, I started using the modules and then forgot to rename the var. Thanks!

**Comment by pcuenca** [comment]
> I'd actually say to not do an automatic displacement to really stay 1-to-1 the same as PyTorch. I really like the fact that you know in PyTorch models are always on "cpu" by default.

That's a breaking change with respect to the current version. No big deal, and it's easy to understand in terms of PyTorch resemblance. I personally liked that you had to do nothing and the pipeline selected the GPU by default; to me, the pipeline is a high-level solution that is there to help you by providing sensible defaults. Similar to disabling gradients computation.

But the alternatives are a bit feeble, so this might well be the best compromise.

**Comment by patrickvonplaten** [comment]
> 

+1 it is indeed a bigger breaking change. 

For me, one of the two options sounds like the best:

1.) 
Remove `torch_device` from `forward(...)` and keep defaulting the init on GPU. Thinking more about it I'm actually fine with it! Also, considering that when passing `device="auto"` to `from_pretrained(...)` of Transformers the model is automatically moved to GPU. I would then maybe add a logging statement though (pipeline moved on GPU...). 
**However** the drawback of this approach is that what do we do when multiple GPUs are available? `device="auto"` in this case moves different layers on different GPUs - that's something that is out of scope here IMO, so we would probably just move it on the first GPU

2.) 
Remove `torch_device` from `forward(...)` and change default init to CPU. Big advantage that we leave the complexity of possible multi-device placement up to the user & it's more PyTorchy. 
Drawback: More of a breaking change, but I think it's fine at that stage of the library.

@patil-suraj @anton-l @pcuenca - let's maybe try to decide somewhat quickly now so that we can include this PR in today's release?

**Comment by anton-l** [comment]
I'm in favor of option 1 out of what Patrick suggested. `"auto"` can be improved a bit later, while defaulting to `"cpu"` might increase friction for first-timers

**Comment by patrickvonplaten** [comment]
> 

Thinking a bit more about it, I'm actually much more for option 2, mainly:

- Don't want to handle the complexity of device placement in `__init__.py`
- People know how PyTorch works -> don't think there is much friction
- Transformer's pipelines also place on CPU by default: https://github.com/huggingface/transformers/blob/e54a1b49aa6268c484625c6374f952f318914743/src/transformers/pipelines/base.py#L750

Happy to adhere to 1. though if @patil-suraj and @pcuenca you prefer

**Comment by pcuenca** [comment]
Those are great points, @patrickvonplaten, let's go for the simpler option and do `2` instead.

**Comment by anton-l** [comment]
Ok, good points, option 2 it is! :)

**Comment by patil-suraj** [comment]
I'm also very much in favor of 2. with `.to` we trying to be more pytorchy, so we should try to mimic as close as possible to avoid any confusion. Let the user handle everything related to device.

**Comment by pcuenca** [comment]
I did `2`, can you please take another look? @patil-suraj @anton-l @patrickvonplaten 

We possibly need to change some documentation and examples. Should we do that in a separate PR?

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 4]
```diff
@@ -29,20 +29,16 @@ def __init__(self, unet, scheduler):
 
     @torch.no_grad()
     def __call__(
-        self, batch_size=1, generator=None, torch_device=None, eta=0.0, num_inference_steps=50, output_type="pil"
+        self, batch_size=1, generator=None, eta=0.0, num_inference_steps=50, output_type="pil"
```
Comment (anton-l) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:32]: `kwargs` here would help with backwards compatibility after we remove `torch_device`

Comment (patrickvonplaten) [reply src/diffusers/pipelines/ddim/pipeline_ddim.py:32]: yes pedro could we do something similar to how we do it in `transformers`? 
E.g.:
```suggestion
        self, batch_size=1, generator=None, eta=0.0, num_inference_steps=50, output_type="pil", **kwargs
        ):
        
        if "torch_device" in kwargs:
            device = kwargs.pop("torch_device")
            warnings.warn(
                "`torch_device` is deprecated as an input argument to `__call__` and will be removed in v0.3.0.
                Consider using `pipe.to(torch_device)` instead."
           )
        # ...set device as previously 
```

Comment (patrickvonplaten) [reply src/diffusers/pipelines/ddim/pipeline_ddim.py:32]: And if `torch_device` is passed we still use it to overwrite (but we can remove it in v0.3.0 already :-))

Comment (patrickvonplaten) [reply src/diffusers/pipelines/ddim/pipeline_ddim.py]: If possible we can apply this to all pipelines

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 5]
```diff
@@ -29,20 +29,25 @@ def __init__(self, unet, scheduler):
...
     def __call__(
-        self, batch_size=1, generator=None, torch_device=None, eta=0.0, num_inference_steps=50, output_type="pil"
+        self, batch_size=1, generator=None, eta=0.0, num_inference_steps=50, output_type="pil", **kwargs
+        ):
+        
+        if "torch_device" in kwargs:
+            device = kwargs.pop("torch_device")
+            warnings.warn(
+                "`torch_device` is deprecated as an input argument to `__call__` and will be removed in v0.3.0.
+                Consider using `pipe.to(torch_device)` instead."
```
Comment (patil-suraj) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:39]:
```suggestion
                "`torch_device` is deprecated as an input argument to `__call__` and will be removed in v0.3.0."
                " Consider using `pipe.to(torch_device)` instead."
```

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 6]
```diff
@@ -29,20 +29,25 @@ def __init__(self, unet, scheduler):
...
+           )
+        # ...set device as previously
```
Comment (patil-suraj) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:41]: (nit) this comment should go above the if cond

**Review by patil-suraj — COMMENTED** [review]
Looking good! Pretty much the same comment as @anton-l and @patrickvonplaten .

Also think the`kwargs` should go in all pipelines.

**Review by anton-l — APPROVED** [review]
Looks great!

**Comment by anton-l** [comment]
I can resolve the conflicts if you'd like, and merge if @patrickvonplaten and @patil-suraj don't have any objections :)

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me! @anton-l feel free to merge!

**Comment by pcuenca** [comment]
> Yes, `kwargs` should be supported in all updated pipelines @pcuenca, sorry for commenting only on DDIM :)

Sure, no problem at all, that's what I understood :)

**Comment by pcuenca** [comment]
Sorry, I have family business going on and can't fix the conflicts right now. Can you do it @anton-l ?, otherwise I'll do it later.
